docs/connections.md

```diff
@@ -0,0 +1,597 @@
+used to type the implicit `this` parameter. 
+
+For example:
+```
+»Foo«.«bar» = 5
+```
+translated to:
+```
+«bar» this:»Foo« = 5
+```
+
+Which is then desugared into a lambda. The introduced name is only `bar`.
+
+If there are any prefix application arguments following the accessor-style
+target, they will be treated as arguments following implicit `this`.
+
+### Overloading
+Only the methods that take `this` as the first parameter can be overloaded. Each
+overload of the given name must have different type of `this`.
+
+However, the type of `this` will be often inferred by the typechecker and it
+IDE cannot tell if given overloads are valid or not.
+
+Example:
+
+```
+«foo» this:«a» = »body1«
+
+«foo» this:«b» = »body2«
+```
+
+In this case `a` and `b` for each `foo` definition will be inferred by the
+compiler. If they end up being different types, overloads are valid. If they are
+the same, an error will be raised.
+
+### Root scope assignments
+
+Any binding in the root scope gets an implicit `this` parameter that describes a
+module. Example:
+```
+main = print "Hello"
+```
+
+`main` here is a function binding that is desugared to lambda. As such its body
+has its own, separate scope. That would not have been a case, if such lines
+appeared in any non-root code block.
+
+## Lambdas
+`arg -> value` is the syntax for lambdas. Left-hand side is a pattern for the
+argument (lambdas are always unary) and the right-hand side is its body.
+Right-hand side, i.e. the lambda body, introduces its own scope.
+
+If the lambda is *not* introduced in what is already a pattern context, the `->`'s pattern introduces identifiers into the scope of the right-hand
+side.
+
+Example:
+```
+«succ» = a -> a + 1
+«foo» = »a«
+```
+
+Here lambda introduces `a` only into its right-hand side. The `a` that is being
+assigned to `foo` is not the same `a` as in lambda — it must be defined
+elsewhere in the module or the code will fail.
+
+However, if `->` appears in a pattern context, its left-hand side identifiers
+are introduced into the scope targeted by the outer pattern context.
+
+Example:
+```
+(a -> b) -> a.»default«
+```
+
+If not for this second rule, the `a` would be visible only in place of
+expression `b`. However, now it is visible in the outer lambda body and can be
+accessed. The only externally provided identifier must be `default` method.
+
+--- 
+Lambdas may not appear in the assignment's pattern (i.e. values cannot be
+pattern-matched into lambda).
+
+So the following is not valid:
+```
+a -> b = foo
+```
+
+---
+Example:
+```
+a -> a -> b
+```
+
+Here the first `a` and second `a` are separate identifiers, the latter shadowing
+the first one. If one wanted to express that both arguments are of the same
+type, `a -> A -> b` would have been used. `b` refers to an identifier from
+graph's scope (it is in the body's position, not pattern).
+
+OPEN QUESTION: actually it might be "nice" to have both `a` unified in such
+case. 
+
+
+
+## Type ascription
```

Comment:
Context ascription (`in`) should also introduce a pattern context.

Reply:
Okay, cool. I'd like to do that consolidation ASAP, as otherwise _those_ docs are likely to evolve without accounting for the information you have here.